aten/src/ATen/core/TransformationHelper.h

```diff
@@ -74,4 +76,59 @@ C10_HOST_DEVICE inline dist_acctype<T> uniform_real(V val, T from, T to) {
   return (x * (to - from) + from);
 }
 
+/**
+ * Transforms normally distributed `val` with mean 0.0 and standard deviation 1.0 to
+ * normally distributed with `mean` and standard deviation `std`.
+ */
+template <typename T>
+C10_HOST_DEVICE inline T normal(T val, T mean, T std) {
+  return val * std + mean;
+}
+
+/**
+ * Transforms uniformly distributed `val` between 0.0 and 1.0 to
+ * Cauchy distribution with location parameter `median` and scale parameter `sigma`.
+ */
+template <typename T>
+C10_HOST_DEVICE inline T cauchy(T val, T median, T sigma) {
+  // https://en.wikipedia.org/wiki/Cauchy_distribution#Cumulative_distribution_function
+  return median + sigma * at::tan(static_cast<T>(M_PI) * (val - static_cast<T>(0.5)));
```

Comment:
I have two questions here. 

(1) This is very minor nit, but instead of the variable name `sigma`, should we use `gamma` or `scale` instead? The variance `sigma**2` is undefined for cauchy. 

(2) I see numpy legacy code for cauchy draws two samples from the normal instead https://github.com/numpy/numpy/blob/main/numpy/random/src/legacy/legacy-distributions.c#L184-L186. My guess is this is because trigonometry `tan` is inefficient? (similar to Polar method, a more efficient version of Box-Muller method with some trigonometry involved). Should we draw two samples from the normal instead as well (I'm not sure how the denominator being 0 (although extremely rare) is handled though) ?

Reply:
re (1) sure send a PR

(2) you're welcome to try this optimization, show it helps, and then send the PR. You will need to handle the zero case somehow though, which might cause trouble

Reply:
OK ~~ thanks @ezyang

Reply:
If we are going to draw 2 samples from normal, that's going to involve computing `sin` and `cos`, as well 2 randoms instead of one, so I doubt it will help, compared to just a single `tan` computation. 
Also, in parallel setting that we care about, polar is *not* more efficient than Box-Muller, even though it avoids computing `sin` and `cos` because it introduces control flow and possibly multiple draws. `sin` and `cos` are really pretty fast on modern GPUs, no reason to avoid them.

Reply:
> If we are going to draw 2 samples from normal, that's going to involve computing sin and cos, as well 2 randoms instead of one, so I doubt it will help, compared to just a single tan computation.

Oh, that's right. This was assuming polar, but looks box-muller is used actually https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/core/DistributionsHelper.h#L180-L187. 

> Also, in parallel setting that we care about, polar is not more efficient than Box-Muller, even though it avoids computing sin and cos because it introduces control flow and possibly multiple draws. sin and cos are really pretty fast on modern GPUs, no reason to avoid them.

Oh, maybe that's why philox randn was also using box-muller. Thanks for the background, this is good to know.